File: surogate/server/routes/conversations.py
"""Conversation API routes – list, detail, turns for chat_turns."""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import delete, func, select, distinct
from sqlalchemy.ext.asyncio import AsyncSession

from surogate.core.db.engine import get_session
from surogate.core.db.models.trace import ChatTurn
from surogate.server.auth.authentication import get_current_subject

router = APIRouter()


@router.get("/")
async def list_conversations(
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
    project_name: Optional[str] = Query(None),
    model: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    limit: int = Query(50, le=200),
    offset: int = Query(0),
):
    """List conversations (grouped from chat_turns)."""
    # Aggregate per conversation_id
    base = select(
        ChatTurn.conversation_id,
        func.min(ChatTurn.created_at).label("started_at"),
        func.max(ChatTurn.created_at).label("last_turn_at"),
        func.count(ChatTurn.id).label("turn_count"),
        func.sum(ChatTurn.prompt_tokens).label("total_prompt_tokens"),
        func.sum(ChatTurn.completion_tokens).label("total_completion_tokens"),
        func.sum(ChatTurn.total_tokens).label("total_tokens"),
        func.avg(ChatTurn.latency_ms).label("avg_latency_ms"),
        func.max(ChatTurn.model).label("model"),
        func.max(ChatTurn.project_name).label("project_name"),
        func.max(ChatTurn.run_name).label("run_name"),
        func.bool_or(ChatTurn.compacted).label("has_compaction"),
    ).group_by(ChatTurn.conversation_id)

    if project_name:
        base = base.where(ChatTurn.project_name == project_name)
    if model:
        base = base.where(ChatTurn.model == model)

    # Wrap as subquery for ordering/pagination
    sub = base.subquery()

    count_stmt = select(func.count()).select_from(sub)
    total = (await session.execute(count_stmt)).scalar() or 0

    rows_stmt = (
        select(sub)
        .order_by(sub.c.last_turn_at.desc())
        .limit(limit)
        .offset(offset)
    )
    rows = (await session.execute(rows_stmt)).all()

    conversations = []
    for r in rows:
        conversations.append({
            "conversation_id": r.conversation_id,
            "started_at": r.started_at.isoformat() if r.started_at else None,
            "last_turn_at": r.last_turn_at.isoformat() if r.last_turn_at else None,
            "turn_count": r.turn_count,
            "total_prompt_tokens": r.total_prompt_tokens or 0,
            "total_completion_tokens": r.total_completion_tokens or 0,
            "total_tokens": r.total_tokens or 0,
            "avg_latency_ms": round(r.avg_latency_ms, 1) if r.avg_latency_ms else None,
            "model": r.model or "",
            "project_name": r.project_name or "",
            "run_name": r.run_name or "",
            "has_compaction": r.has_compaction or False,
        })

    return {"conversations": conversations, "total": total}


@router.get("/{conversation_id}")
async def get_conversation(
    conversation_id: str,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Get a single conversation with all its turns."""
    stmt = (
        select(ChatTurn)
        .where(ChatTurn.conversation_id == conversation_id)
        .order_by(ChatTurn.created_at.asc())
    )
    result = await session.execute(stmt)
    turns = result.scalars().all()

    if not turns:
        raise HTTPException(status_code=404, detail="Conversation not found")

    first = turns[0]
    last = turns[-1]

    total_prompt = sum(t.prompt_tokens or 0 for t in turns)
    total_completion = sum(t.completion_tokens or 0 for t in turns)
    latencies = [t.latency_ms for t in turns if t.latency_ms is not None]
    avg_latency = round(sum(latencies) / len(latencies), 1) if latencies else None

    # Extract messages from turns for the thread view
    messages = []
    for t in turns:
        req = t.request_body or {}
        turn_messages = req.get("messages", [])
        # Only add the last user message (the prompt for this turn)
        if turn_messages:
            last_msg = turn_messages[-1]
            if last_msg.get("role") == "user":
                messages.append({
                    "role": "user",
                    "content": last_msg.get("content", ""),
                    "timestamp": t.created_at.isoformat() if t.created_at else None,
                    "tokens": t.prompt_tokens,
                })

        # Add assistant response from response_body
        resp = t.response_body
        if resp:
            choices = resp.get("choices", [])
            if choices:
                assistant_msg = choices[0].get("message", {})
                content = assistant_msg.get("content", "")
                reasoning = assistant_msg.get("reasoning_content", "")
                messages.append({
                    "role": "assistant",
                    "content": content,
                    "reasoning_content": reasoning or None,
                    "timestamp": t.created_at.isoformat() if t.created_at else None,
                    "tokens": t.completion_tokens,
                    "latency": round(t.latency_ms) if t.latency_ms else None,
                })

    turn_details = []
    for t in turns:
        turn_details.append({
            "id": t.id,
            "parent_hash": t.parent_hash,
            "state_hash": t.state_hash,
            "tail_hash": t.tail_hash,
            "model": t.model,
            "is_streaming": t.is_streaming,
            "prompt_tokens": t.prompt_tokens,
            "completion_tokens": t.completion_tokens,
            "total_tokens": t.total_tokens,
            "latency_ms": t.latency_ms,
            "compacted": t.compacted,
            "created_at": t.created_at.isoformat() if t.created_at else None,
        })

    # Extract system prompt from the first turn's request
    system_prompt = None
    first_req = first.request_body or {}
    for msg in first_req.get("messages", []):
        if msg.get("role") == "system":
            system_prompt = msg.get("content", "")
            break

    return {
        "conversation_id": conversation_id,
        "project_name": first.project_name,
        "run_name": first.run_name,
        "model": last.model,
        "system_prompt": system_prompt,
        "started_at": first.created_at.isoformat() if first.created_at else None,
        "last_turn_at": last.created_at.isoformat() if last.created_at else None,
        "turn_count": len(turns),
        "total_prompt_tokens": total_prompt,
        "total_completion_tokens": total_completion,
        "total_tokens": total_prompt + total_completion,
        "avg_latency_ms": avg_latency,
        "has_compaction": any(t.compacted for t in turns),
        "messages": messages,
        "turns": turn_details,
    }


@router.delete("/{conversation_id}")
async def delete_conversation(
    conversation_id: str,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Delete all turns belonging to a conversation."""
    stmt = delete(ChatTurn).where(ChatTurn.conversation_id == conversation_id)
    result = await session.execute(stmt)
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Conversation not found")
    await session.commit()
    return {"status": "deleted", "turns_deleted": result.rowcount}

File: frontend/src/api/conversations.ts
// Copyright (c) 2026, Invergent SA, developed by Flavius Burca
// SPDX-License-Identifier: AGPL-3.0-only
//
import { authFetch } from "@/api/auth";
import type { Conversation, ConversationDetail, ConversationTurn, ConversationMessage } from "@/features/conversations/conversations-data";

// -- Raw backend types (snake_case) ----------------------------------------

interface RawConversationSummary {
  conversation_id: string;
  started_at: string | null;
  last_turn_at: string | null;
  turn_count: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  total_tokens: number;
  avg_latency_ms: number | null;
  model: string;
  project_name: string;
  run_name: string;
  has_compaction: boolean;
}

interface RawConversationListResponse {
  conversations: RawConversationSummary[];
  total: number;
}

interface RawTurn {
  id: string;
  parent_hash: string | null;
  state_hash: string;
  tail_hash: string;
  model: string;
  is_streaming: boolean;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  latency_ms: number | null;
  compacted: boolean;
  created_at: string | null;
}

interface RawMessage {
  role: "user" | "assistant";
  content: string;
  reasoning_content?: string | null;
  timestamp: string | null;
  tokens: number | null;
  latency: number | null;
}

interface RawConversationDetail {
  conversation_id: string;
  project_name: string;
  run_name: string;
  model: string;
  system_prompt: string | null;
  started_at: string | null;
  last_turn_at: string | null;
  turn_count: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  total_tokens: number;
  avg_latency_ms: number | null;
  has_compaction: boolean;
  messages: RawMessage[];
  turns: RawTurn[];
}

// -- Response types --------------------------------------------------------

export interface ConversationListResponse {
  conversations: Conversation[];
  total: number;
}

// -- Transform -------------------------------------------------------------

function formatRelativeTime(iso: string | null): string {
  if (!iso) return "";
  const diff = Date.now() - new Date(iso).getTime();
  const mins = Math.floor(diff / 60000);
  if (mins < 1) return "now";
  if (mins < 60) return `${mins}m ago`;
  const hours = Math.floor(mins / 60);
  if (hours < 24) return `${hours}h ago`;
  const days = Math.floor(hours / 24);
  return `${days}d ago`;
}

function formatDuration(startIso: string | null, endIso: string | null): string {
  if (!startIso || !endIso) return "";
  const ms = new Date(endIso).getTime() - new Date(startIso).getTime();
  const secs = Math.floor(ms / 1000);
  if (secs < 60) return `${secs}s`;
  const mins = Math.floor(secs / 60);
  const remSecs = secs % 60;
  return `${mins}m ${remSecs}s`;
}

function transformSummary(r: RawConversationSummary): Conversation {
  return {
    id: r.conversation_id,
    model: r.model,
    projectName: r.project_name,
    runName: r.run_name,
    turnCount: r.turn_count,
    tokensIn: r.total_prompt_tokens,
    tokensOut: r.total_completion_tokens,
    totalTokens: r.total_tokens,
    avgLatencyMs: r.avg_latency_ms,
    startedAt: r.started_at,
    lastTurnAt: r.last_turn_at,
    hasCompaction: r.has_compaction,
    time: formatRelativeTime(r.last_turn_at),
    duration: formatDuration(r.started_at, r.last_turn_at),
  };
}

function transformDetail(r: RawConversationDetail): ConversationDetail {
  return {
    id: r.conversation_id,
    model: r.model,
    projectName: r.project_name,
    runName: r.run_name,
    turnCount: r.turn_count,
    tokensIn: r.total_prompt_tokens,
    tokensOut: r.total_completion_tokens,
    totalTokens: r.total_tokens,
    avgLatencyMs: r.avg_latency_ms,
    startedAt: r.started_at,
    lastTurnAt: r.last_turn_at,
    hasCompaction: r.has_compaction,
    time: formatRelativeTime(r.last_turn_at),
    duration: formatDuration(r.started_at, r.last_turn_at),
    systemPrompt: r.system_prompt,
    messages: r.messages.map((m) => ({
      role: m.role,
      content: m.content,
      reasoningContent: m.reasoning_content ?? undefined,
      timestamp: m.timestamp,
      tokens: m.tokens ?? 0,
      latency: m.latency ?? undefined,
    })),
    turns: r.turns.map((t) => ({
      id: t.id,
      parentHash: t.parent_hash,
      stateHash: t.state_hash,
      tailHash: t.tail_hash,
      model: t.model,
      isStreaming: t.is_streaming,
      promptTokens: t.prompt_tokens,
      completionTokens: t.completion_tokens,
      totalTokens: t.total_tokens,
      latencyMs: t.latency_ms,
      compacted: t.compacted,
      createdAt: t.created_at,
    })),
  };
}

// -- API functions ---------------------------------------------------------

export async function listConversations(params?: {
  project_name?: string;
  model?: string;
  search?: string;
  limit?: number;
  offset?: number;
}): Promise<ConversationListResponse> {
  const qs = new URLSearchParams();
  if (params?.project_name) qs.set("project_name", params.project_name);
  if (params?.model) qs.set("model", params.model);
  if (params?.search) qs.set("search", params.search);
  if (params?.limit) qs.set("limit", String(params.limit));
  if (params?.offset) qs.set("offset", String(params.offset));

  const url = `/api/conversations/${qs.toString() ? `?${qs}` : ""}`;
  const response = await authFetch(url);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to list conversations");
  }
  const raw = (await response.json()) as RawConversationListResponse;
  return {
    conversations: raw.conversations.map(transformSummary),
    total: raw.total,
  };
}

export async function getConversation(conversationId: string): Promise<ConversationDetail> {
  const response = await authFetch(`/api/conversations/${conversationId}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to get conversation");
  }
  const raw = (await response.json()) as RawConversationDetail;
  return transformDetail(raw);
}

export async function deleteConversation(conversationId: string): Promise<void> {
  const response = await authFetch(`/api/conversations/${conversationId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to delete conversation");
  }
}
